File: 0051-n-queens/N-Queens.swift
extension Solution {
    func solveNQueens(_ n: Int) -> [[String]] {
        var m1 = [Bool](repeating: false, count: n)  // row
        var m2 = [Bool](repeating: false, count: n)  // column
        var m3 = [Bool](repeating: false, count: n * 2)  // slash
        var m4 = [Bool](repeating: false, count: n * 2)  // backslash
        var board = [[Bool]](repeating: [Bool](repeating: false, count: n), count: n)

        var rv: [[String]] = []
        func solve(_ y: Int) {
            if y >= n {
                var ss = [String]()
                for i in 0..<n {
                    var cs = [Character](repeating: ".", count: n)
                    for j in 0..<n {
                        if board[i][j] {
                            cs[j] = "Q"
                        }
                    }
                    ss.append(String(cs))
                }
                rv.append(ss)
                return
            }
            for x in 0..<n {
                if m1[x] || m2[y] || m3[x + y] || m4[n + x - y] {
                    continue
                }
                board[x][y] = true
                (m1[x], m2[y], m3[x + y], m4[n + x - y]) = (true, true, true, true)
                solve(y + 1)
                (m1[x], m2[y], m3[x + y], m4[n + x - y]) = (false, false, false, false)
                board[x][y] = false
            }
        }
        solve(0)
        return rv
    }
}

File: 0000-0099/0051-n-queens/n-queens.go
package solution

func solveNQueens(n int) [][]string {
	var (
		m1 []bool = make([]bool, n)   // column
		m2 []bool = make([]bool, n)   // rows
		m3 []bool = make([]bool, n*2) // slash
		m4 []bool = make([]bool, n*2) // backslash
	)
	var board [][]bool = make([][]bool, n)
	for i := 0; i < n; i++ {
		board[i] = make([]bool, n)
	}

	var ret [][]string

	var solve func(int)
	solve = func(y int) {
		if y >= n {
			rv := make([]string, n)
			for i := 0; i < n; i++ {
				for j := 0; j < n; j++ {
					if board[j][i] {
						rv[i] += "Q"
					} else {
						rv[i] += "."
					}
				}
			}
			ret = append(ret, rv)
			return
		}
		for x := 0; x < n; x++ {
			if m1[x] || m2[y] || m3[x+y] || m4[n+x-y] {
				continue
			}
			board[x][y] = true
			m1[x], m2[y], m3[x+y], m4[n+x-y] = true, true, true, true
			solve(y + 1)
			board[x][y] = false
			m1[x], m2[y], m3[x+y], m4[n+x-y] = false, false, false, false
		}
	}

	solve(0)

	return ret
}
